### PR TITLE
feat: Add `WildcardCharacter`, `WildcardExpression`, and `WildcardExpressionView` to process wildcard patterns. 

### PR DESCRIPTION
**Pull request by SharafMohamed**
# Description
This PR introduces three core classes to represent and manipulate wildcard search patterns efficiently:
- `Wildcard Character` - Represents a single character in an expression, including its type (i.e., normal, greedy wildcard `*`, non-greedy wildcard `?`, or escape `\`).
- `WildcardExpression` - Owns a processed wildcard pattern string and a vector of `WildcardCharacter`s.
- `WildcardExpressionView` - A lightweight, non-owning view of `WildcardExpression`. It uses `std::span` and std::string_view` to avoid copies and ensure bounds safety. It also offers utilities for generating a regex equivalent of the subrange, checking for boundary wildcards, extending views, and validating well-formedness relative to escaped characters.

# Validation
-

### CODE REVIEW

File: src/log_surgeon/wildcard_query_parser/WildcardExpressionView.cpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    auto [begin_idx, end_idx]{get_indicies()};
+
+    std::span const full_span{m_expression->get_chars()};
+
+    if (begin_idx > 0 && full_span[begin_idx - 1].is_greedy_wildcard()) {
+        --begin_idx;
+    }
+    if (end_idx < full_span.size() && full_span[end_idx].is_greedy_wildcard()) {
+        ++end_idx;
+    }
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply src/log_surgeon/wildcard_query_parser/WildcardExpressionView.cpp:37]: This won't be useful for the purposed of this method when we use it for search.

File: src/log_surgeon/wildcard_query_parser/WildcardExpressionView.cpp [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            regex_string += ".";
+        } else if (SchemaParser::get_special_regex_characters().contains(value)) {
+            regex_string += "\\";
+            regex_string += value;
+        } else {
+            regex_string += value;
+        }
+    }
+    return regex_string;
+}
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply src/log_surgeon/wildcard_query_parser/WildcardExpressionView.cpp:77]: The well_formed check is to be used by the caller. A regex string can still be made for a malformed string if truly desired.

File: src/log_surgeon/wildcard_query_parser/WIldcardCharacter.hpp [hunk 3]
```diff
@@ -0,0 +1,36 @@
+#ifndef LOG_SURGEON_QUERY_PARSER_WILDCARD_CHARACTER_HPP
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply src/log_surgeon/wildcard_query_parser/WIldcardCharacter.hpp]: Fixed.

File: src/log_surgeon/wildcard_query_parser/WildcardExpressionView.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply src/log_surgeon/wildcard_query_parser/WildcardExpressionView.cpp]: Added.

File: src/log_surgeon/wildcard_query_parser/WildcardExpressionView.hpp [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+ */
+class WildcardExpressionView {
+public:
+    WildcardExpressionView(WildcardExpression const& expression, size_t begin_idx, size_t end_idx);
+
+    /**
+     * @return A copy of this view, but extended to include adjacent greedy wildcards.
+     */
+    [[nodiscard]] auto extend_to_adjacent_greedy_wildcards() const -> WildcardExpressionView;
+
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply src/log_surgeon/wildcard_query_parser/WildcardExpressionView.hpp:33]: Its fine as is.

**Review by davidlion — CHANGES_REQUESTED** [review]

File: src/log_surgeon/wildcard_query_parser/WildcardExpression.hpp
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply src/log_surgeon/wildcard_query_parser/WildcardExpression.hpp:1]: `m_processed_search_string` is naming left over from CLP. I don't think it needs to be specially processed in any way as far as this class is concerned. The caller is the one that may want to process the string for efficiency.

File: src/log_surgeon/wildcard_query_parser/ExpressionCharacter.hpp [hunk 5]
```diff
@@ -0,0 +1,34 @@
+#ifndef LOG_SURGEON_QUERY_PARSER_EXPRESSION_CHARACTER_HPP
+#define LOG_SURGEON_QUERY_PARSER_EXPRESSION_CHARACTER_HPP
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply src/log_surgeon/wildcard_query_parser/ExpressionCharacter.hpp:2]: Actually all these should be renamed to use WILDCARD_QUERY_PARSER based on the file name, and the redundant wildcard prefix should probably be dropped from all classes.

File: tests/test-wildcard-expression.cpp [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+    auto const& expression_chars{expression.get_chars()};
+    REQUIRE(input.size() == expression_chars.size());
+
+    for (auto const& expression_char : expression_chars) {
+        REQUIRE_FALSE(expression_char.is_greedy_wildcard());
+        REQUIRE_FALSE(expression_char.is_non_greedy_wildcard());
+        REQUIRE_FALSE(expression_char.is_escape());
+    }
+}
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply tests/test-wildcard-expression.cpp]: Changed.

File: src/log_surgeon/wildcard_query_parser/ExpressionCharacter.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply src/log_surgeon/wildcard_query_parser/ExpressionCharacter.hpp]: Changed, can you double check i didn't miss the change anywhere?

File: src/log_surgeon/wildcard_query_parser/Expression.cpp [hunk 7]
```diff
@@ -0,0 +1,26 @@
+#include "Expression.hpp"
+
+#include <string>
+#include <utility>
+
+#include <log_surgeon/wildcard_query_parser/ExpressionCharacter.hpp>
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply src/log_surgeon/wildcard_query_parser/Expression.cpp:6]: This explicitly refers to ExpressionCharacter in the cpp, therefore the include should be there.

File: src/log_surgeon/wildcard_query_parser/ExpressionView.cpp [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+#include <algorithm>
+#include <cstddef>
+#include <span>
+#include <string>
+#include <string_view>
+#include <utility>
+
+#include <log_surgeon/SchemaParser.hpp>
+#include <log_surgeon/wildcard_query_parser/Expression.hpp>
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply src/log_surgeon/wildcard_query_parser/ExpressionView.cpp:11]: Same as above, explicitly used in cpp.

File: src/log_surgeon/wildcard_query_parser/ExpressionView.hpp [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+    [[nodiscard]] auto get_search_string() const -> std::string_view { return m_search_string; }
+
+private:
+    [[nodiscard]] auto get_indices() const -> std::pair<size_t, size_t> {
+        auto const& full_chars{m_expression->get_chars()};
+        auto const* begin_ptr{m_chars.data()};
+        auto begin_idx{static_cast<size_t>(begin_ptr - full_chars.data())};
+        return {begin_idx, begin_idx + m_chars.size()};
+    }
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply src/log_surgeon/wildcard_query_parser/ExpressionView.hpp:84]: Its fine.

File: tests/test-expression-view.cpp [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+#include <log_surgeon/wildcard_query_parser/ExpressionView.hpp>
+
+#include <catch2/catch_test_macros.hpp>
+
+/**
+ * @defgroup unit_tests_expression_view `ExpressionView` unit tests.
+ * @brief Unit tests for `ExpressionView` to ... .
+
+ * These unit tests contain the `ExpressionView` tag.
+ */
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply tests/test-expression-view.cpp:15]: Will update when all tests are added.

File: src/log_surgeon/wildcard_query_parser/ExpressionView.cpp [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    ExpressionView wildcard_expression_view{*m_expression, begin_idx, end_idx};
+    return {is_extended, wildcard_expression_view};
+}
+
+auto ExpressionView::is_well_formed() const -> bool {
+    if (m_chars.empty()) {
+        return true;
+    }
+    auto const [begin_idx, end_idx]{get_indices()};
+    if (begin_idx > 0 && m_expression->get_chars()[begin_idx - 1].is_escape()) {
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply src/log_surgeon/wildcard_query_parser/ExpressionView.cpp:50]: Its fine as is.

File: tests/test-expression.cpp [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+TEST_CASE("empty_expression", "[Expression]") {
+    Expression const expression{""};
+    REQUIRE(expression.get_search_string().empty());
+    REQUIRE(expression.get_chars().empty());
+}
+
+/**
+ * @ingroup unit_tests_expression
+ * @brief Tests a `Expression` with only normal characters.
+ */
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply tests/test-expression.cpp]: Fixed.

File: tests/test-expression.cpp [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+            REQUIRE(expression_char.is_greedy_wildcard());
+            REQUIRE_FALSE(expression_char.is_non_greedy_wildcard());
+            REQUIRE_FALSE(expression_char.is_escape());
+        } else {
+            REQUIRE_FALSE(expression_char.is_greedy_wildcard());
+            REQUIRE(expression_char.is_non_greedy_wildcard());
+            REQUIRE_FALSE(expression_char.is_escape());
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply tests/test-expression.cpp:80]: The escaped wildcard is already handled (and is also not what you show). The other cases you mention aren't really edge cases, the current tests should be enough to ensure they also work.

File: tests/test-expression.cpp [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+            REQUIRE_FALSE(expression_char.is_greedy_wildcard());
+            REQUIRE_FALSE(expression_char.is_non_greedy_wildcard());
+            REQUIRE(expression_char.is_escape());
+        } else {
+            REQUIRE_FALSE(expression_char.is_greedy_wildcard());
+            REQUIRE_FALSE(expression_char.is_non_greedy_wildcard());
+            REQUIRE_FALSE(expression_char.is_escape());
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply tests/test-expression.cpp:111]: This is pretty much covered in the existing case.

[the remark this answers is not in the document]
Comment (SharafMohamed) [reply tests/test-expression.cpp:111]: Are those grammar improvements not made already?

File: tests/test-expression.cpp [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+            REQUIRE(expression_char.is_greedy_wildcard());
+            REQUIRE_FALSE(expression_char.is_non_greedy_wildcard());
+            REQUIRE_FALSE(expression_char.is_escape());
+        } else {
+            REQUIRE_FALSE(expression_char.is_greedy_wildcard());
+            REQUIRE(expression_char.is_non_greedy_wildcard());
+            REQUIRE_FALSE(expression_char.is_escape());
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply tests/test-expression.cpp]: Changed.

File: tests/test-expression-view.cpp [hunk 16]
```diff
@@ -0,0 +1,283 @@
+#include <cstddef>
+#include <string>
+
+#include <log_surgeon/wildcard_query_parser/Expression.hpp>
+#include <log_surgeon/wildcard_query_parser/ExpressionView.hpp>
+
+#include <catch2/catch_message.hpp>
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply tests/test-expression-view.cpp:7]: clang-tidy complains without this header.

File: tests/test-expression-view.cpp [hunk 17]
```diff
@@ -0,0 +1,283 @@
...
+    REQUIRE_FALSE(is_extended);
+    REQUIRE(view.get_search_string() == extended_view.get_search_string());
+}
+
+/**
+ * @ingroup unit_tests_expression_view
+ * @brief Tests bound snapping during `ExpressionView`construction.
+ *
+ * Negative casted values test wrap-around behavior.
+ */
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply tests/test-expression-view.cpp]: Modified.

File: tests/test-expression-view.cpp [hunk 18]
```diff
@@ -0,0 +1,343 @@
...
+    SECTION("empty_subrange") {
+        string const input{"abc"};
+        Expression const expression{input};
+        ExpressionView const view{expression, 0, 0};
+
+        auto const [regex_string, contains_wildcard]{view.generate_regex_string()};
+        REQUIRE(regex_string.empty());
+        REQUIRE_FALSE(contains_wildcard);
+    }
+}
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply tests/test-expression-view.cpp]: Fixed.

File: tests/test-expression-view.cpp [hunk 19]
```diff
@@ -0,0 +1,343 @@
...
+
+#include <log_surgeon/wildcard_query_parser/Expression.hpp>
+#include <log_surgeon/wildcard_query_parser/ExpressionView.hpp>
+
+#include <catch2/catch_message.hpp>
+#include <catch2/catch_test_macros.hpp>
+
+/**
+ * @defgroup unit_tests_expression_view `ExpressionView` unit tests.
+ * @brief Unit tests for `ExpressionView` to ... .
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply tests/test-expression-view.cpp]: Whoops, updated. Also updated the expression unit-test brief.

**Review by davidlion — DISMISSED** [review]

File: tests/test-expression-view.cpp [hunk 20]
```diff
@@ -0,0 +1,344 @@
...
+    SECTION("empty_subrange") {
+        string const input{"abc"};
+        Expression const expression{input};
+        ExpressionView const view{expression, 0, 0};
+
+        auto const [regex_string, contains_wildcard]{view.generate_regex_string()};
+        REQUIRE(regex_string.empty());
+        REQUIRE_FALSE(contains_wildcard);
+    }
+}
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply tests/test-expression-view.cpp]: Added.

File: tests/test-expression-view.cpp [hunk 21]
```diff
@@ -0,0 +1,355 @@
...
+            if (start_pos == cEscapePos1 + 1 || start_pos == cEscapePos2 + 1) {
+                REQUIRE_FALSE(view.is_well_formed());
+            } else if (end_pos == cEscapePos1 + 1 || end_pos == cEscapePos2 + 1) {
+                REQUIRE_FALSE(view.is_well_formed());
+            } else {
+                REQUIRE(view.is_well_formed());
+            }
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply tests/test-expression-view.cpp:200]: But the substrings do end in a trailing backslash
